Storage/Storage/CoreData/IterativeStoreMigrator.swift

```diff
@@ -0,0 +1,220 @@
+    ///
+    /// - Parameters:
+    ///     - sourceStore: URL of the store on disk.
+    ///     - storeType: Type of store (usually NSSQLiteStoreType).
+    ///     - to: The target/most current model the migrator should migrate to.
+    ///     - using: List of models on disk, sorted in migration order, that should include the to: model.
+    ///
+    /// - Throws: A whole bunch of crap is possible to be thrown between Core Data and FileManager.
+    ///
+    static func iterativeMigrate(sourceStore: URL, storeType: String, to targetModel:NSManagedObjectModel, using modelNames: [String]) throws -> Bool {
+        // If the persistent store does not exist at the given URL,
+        // assume that it hasn't yet been created and return success immediately.
+        guard FileManager.default.fileExists(atPath: sourceStore.path) == true else {
+            return true
+        }
+
+        // Get the persistent store's metadata.  The metadata is used to
+        // get information about the store's managed object model.
+        guard let sourceMetadata = try metadataForPersistentStore(storeType: storeType, at: sourceStore) else {
+            return false
+        }
+
+        // Check whether the final model is already compatible with the store.
+        // If it is, no migration is necessary.
+        guard targetModel.isConfiguration(withName: nil, compatibleWithStoreMetadata: sourceMetadata) == false else {
+            return true
+        }
+
+        // Find the current model used by the store.
+        guard let sourceModel = try model(for: sourceMetadata) else {
+            return false
+        }
+
+        // Get NSManagedObjectModels for each of the model names given.
+        let objectModels = try models(for: modelNames)
+
+        // Build an inclusive list of models between the source and final models.
+        var modelsToMigrate = [NSManagedObjectModel]()
+        var firstFound = false, lastFound = false, reverse = false
+
+        for model in objectModels {
+            if model.isEqual(sourceModel) || model.isEqual(targetModel) {
+                if firstFound {
+                    lastFound = true
+                    // In case a reverse migration is being performed (descending through the
+                    // ordered array of models), check whether the source model is found
+                    // after the final model.
+                    reverse = model.isEqual(sourceModel)
+                } else {
+                    firstFound = true
+                }
+            }
+
+            if firstFound {
+                modelsToMigrate.append(model)
+            }
+
+            if lastFound {
+                break
+            }
+        }
+
+        // Ensure that the source model is at the start of the list.
+        if reverse {
+            modelsToMigrate = modelsToMigrate.reversed()
+        }
+
+        // Migrate through the list
+        for index in 0...(modelsToMigrate.count - 2) {
```

Comment:
Tiny nit: As a note to our future selves it might be worthwhile to expound on this comment to explain what is happening here (e.g. why `count - 2`).